### DISCUSSION TITLE
CLI tool(s) to flash .hex files in windows ?

### DISCUSSION THREAD

**Discussion by Josverl** [micropython/micropython#10934] [STM32 / Pyboard] [answered]
Today I though i finally managed to fry my trusty Pybd1 as after flashing a new firmware using `dfu-util 0.11`
all files on the filesystem contained nothing but 0xFF.
Note that the firmware still booted, just found it hard to deal with unreadable boot.py and main.py. 
Also new files can be written no problem.

 * Tried using older firmwares, and using the scripts I've used before - same issue 
 * Installed stm32cubeprogrammer - to find that it cannot deal with .dfu files
   * downloaded the '.hex' files
   * flashed through the GUI
   * success, the board is alive again 😸

Now while the GUI is nice , I'd like to find a CLI tool that can do the same from the commandline 
**So the question are :** 
 - Is there 'a CLI version of STM32CubeProgrammer' that handles `.hex` files and runs on Windows (or Python) 
 - is there a way to extract/convert `.dfu` files to `.hex` files ?
 

**Setup:** (so many 11's ...)  
 - Windows 11 (10.0.22621) 
 - pybd 1.1 
 - dfu-util 0.11
 - CubeProgrammer v2.13.0 Win x64

**Comment by jimmo** [comment, marked as answer]
> Is there 'a CLI version of STM32CubeProgrammer' that handles .hex files and runs on Windows (or Python)

@Josverl Yes! STM32CubeProgrammer :)

See section 3 of https://www.st.com/resource/en/user_manual/um2237-stm32cubeprogrammer-software-description-stmicroelectronics.pdf

**Reply by Josverl** [reply]
Thanks, perhaps time for a rewrite of Pythons at the Windows Zoo ...

I'll give it a spin and see if I can do a short writeup

**Reply by Josverl** [reply]
Also,
 is there a way to extract/convert .dfu files to .hex files ?

As the GUI cannot digest .DFU files, I don't expect any different from the cli

**Reply by jimmo** [reply]
> Thanks, perhaps time for a rewrite of Pythons at the Windows Zoo ...

That would make me very happy.

> is there a way to extract/convert .dfu files to .hex files ?

It is baffling to me why CubeProgrammer doesn't support .dfu files. I'd love to know ST's reasoning behind this.

DfuSe does have a converter in it. https://www.st.com/en/development-tools/stsw-stm32080.html

We also have a tool in tools/dfu.py that can convert from .dfu to .bin, which you can use from Cube Programmer. (see the --dump option).
